File: Sources/PiInfobar/StatsEngine.swift
import Foundation
import SwiftUI

@MainActor
final class StatsEngine: ObservableObject {

    @Published var loading = true
    @Published var progress = 0.0
    @Published var error: String?
    @Published var lastUpdated: Date?

    /// The parsed per-day aggregates (full history).
    private var days: [DayAgg] = []

    nonisolated static var sessionsDir: URL {
        let home = FileManager.default.homeDirectoryForCurrentUser
        if UserDefaults.standard.bool(forKey: SettingsKeys.remoteSyncEnabled) {
            return home.appendingPathComponent(".pi/remote-agent-sessions")
        }
        return home.appendingPathComponent(".pi/agent/sessions")
    }

    nonisolated static var cacheURL: URL {
        let home = FileManager.default.homeDirectoryForCurrentUser
        if UserDefaults.standard.bool(forKey: SettingsKeys.remoteSyncEnabled) {
            return home.appendingPathComponent(".pi/pi-infobar-remote-cache.json")
        }
        return home.appendingPathComponent(".pi/pi-infobar-cache.json")
    }

    // MARK: - Public API

    nonisolated static func performRemoteSync() async throws {
        let defaults = UserDefaults.standard
        let host = defaults.string(forKey: SettingsKeys.remoteHost) ?? ""
        let port = defaults.string(forKey: SettingsKeys.remotePort) ?? "22"
        let user = defaults.string(forKey: SettingsKeys.remoteUser) ?? ""
        let keyPath = defaults.string(forKey: SettingsKeys.remoteKeyPath) ?? "~/.ssh/id_rsa"
        let remotePath = defaults.string(forKey: SettingsKeys.remotePath) ?? "~/.pi/agent/sessions"

        guard !host.isEmpty, !user.isEmpty else {
            throw NSError(domain: "PiStats", code: 1, userInfo: [NSLocalizedDescriptionKey: "Remote host and username must be configured in Settings."])
        }

        try await RemoteSync.sync(
            host: host,
            port: port,
            user: user,
            keyPath: keyPath,
            remotePath: remotePath,
            localPath: sessionsDir
        )
    }

    func load(force: Bool = false) {
        loading = true
        progress = 0
        error = nil
        Task.detached(priority: .userInitiated) {
            do {
                if UserDefaults.standard.bool(forKey: SettingsKeys.remoteSyncEnabled) {
                    do {
                        try await Self.performRemoteSync()
                    } catch {
                        let cacheExists = FileManager.default.fileExists(atPath: Self.cacheURL.path)
                        if !cacheExists {
                            throw error
                        }
                        NSLog("Remote sync failed: \(error.localizedDescription). Using cached data.")
                    }
                }

                let agg = try Self.buildAggregate(force: force) { p in
                    Task { @MainActor in self.progress = p }
                }
                await MainActor.run {
                    self.days = agg.days
                    self.lastUpdated = agg.generatedAt
                    self.loading = false
                    self.progress = 1
                }
            } catch {
                await MainActor.run {
                    self.error = error.localizedDescription
                    self.loading = false
                }
            }
        }
    }

    /// Build a summary for the given range from already-parsed days.
    func summary(for range: TimeRange) -> StatsSummary {
        Self.summarize(days: days, range: range)
    }

    private var todayAgg: DayAgg? {
        let key = Self.dayFormatter.string(from: Date())
        return days.first(where: { $0.date == key })
    }

    /// Today's cost — used for the menu bar title.
    var todayCost: Double { todayAgg?.cost ?? 0 }
    var todayLines: Int { todayAgg?.langLines.values.reduce(0, +) ?? 0 }
    var todayMessages: Int { (todayAgg.map { $0.userMsgs + $0.asstMsgs }) ?? 0 }
    var todaySessions: Int { todayAgg?.sessionIds.count ?? 0 }
    var totalCostAll: Double { days.reduce(0) { $0 + $1.cost } }

    // MARK: - Aggregation / summarizing

    nonisolated static func summarize(days: [DayAgg], range: TimeRange) -> StatsSummary {
        let cal = Calendar.current
        let todayKey = dayFormatter.string(from: Date())

        // Filter days by range.
        let filtered: [DayAgg]
        if let nDays = range.days {
            let cutoff = cal.date(byAdding: .day, value: -(nDays - 1), to: cal.startOfDay(for: Date()))!
            filtered = days.filter { d in
                guard let dt = dayFormatter.date(from: d.date) else { return false }
                return dt >= cutoff
            }
        } else {
            filtered = days
        }

        var s = StatsSummary()
        var sessionSet = Set<String>()
        var langLines: [String: Int] = [:]
        var langEdits: [String: Int] = [:]
        var modelCost: [String: Double] = [:]
        var modelCount: [String: Int] = [:]
        var projCost: [String: Double] = [:]
        var projSessions: [String: Set<String>] = [:]
        var toolCount: [String: Int] = [:]
        var spend: [DaySpend] = []

        for d in filtered {
            s.totalCost += d.cost
            s.inTok += d.inTok; s.outTok += d.outTok
            s.crTok += d.crTok; s.cwTok += d.cwTok
            s.userMsgs += d.userMsgs; s.asstMsgs += d.asstMsgs
            s.toolResults += d.toolResults
            sessionSet.formUnion(d.sessionIds)
            for (k, v) in d.langLines { langLines[k, default: 0] += v }
            for (k, v) in d.langEdits { langEdits[k, default: 0] += v }
            for (k, v) in d.modelCost { modelCost[k, default: 0] += v }
            for (k, v) in d.modelCount { modelCount[k, default: 0] += v }
            for (k, v) in d.projectCost { projCost[k, default: 0] += v }
            for (k, ids) in d.projectSessions { projSessions[k, default: []].formUnion(ids) }
            for (k, v) in d.toolCount { toolCount[k, default: 0] += v }
            if d.date == todayKey { s.todayCost += d.cost }
            if let dt = dayFormatter.date(from: d.date) {
                spend.append(DaySpend(date: d.date, day: dt, cost: d.cost))
            }
        }

        s.sessionCount = sessionSet.count
        s.daysActive = filtered.filter { $0.cost > 0 || $0.asstMsgs > 0 }.count
        s.avgCostPerDay = s.daysActive > 0 ? s.totalCost / Double(s.daysActive) : 0

        // Languages
        s.languages = langLines.map { (lang, lines) in
            let info = LanguageMap.infoFor(lang)
            return LangStat(name: lang, lines: lines, edits: langEdits[lang] ?? 0,
                            color: info.color, symbol: info.symbol)
        }
        .filter { $0.edits > 0 }
        .sorted { $0.lines == $1.lines ? $0.edits > $1.edits : $0.lines > $1.lines }

        // Models
        s.models = modelCount.map { (m, c) in
            ModelStat(name: m, displayName: prettyModel(m), cost: modelCost[m] ?? 0,
                      count: c, color: modelColor(m))
        }
        .sorted { $0.cost > $1.cost }

        // Projects
        s.projects = projCost.map { (p, c) in
            ProjectStat(name: p, cost: c, sessions: projSessions[p]?.count ?? 0)
        }
        .sorted { $0.cost > $1.cost }

        // Tools
        s.tools = toolCount.map { ToolStat(name: $0.key, count: $0.value) }
            .sorted { $0.count > $1.count }

        // Daily spend (sorted ascending, fill gaps so the chart looks continuous)
        var filledSpend: [DaySpend] = []
        let startDate: Date
        let endDate = cal.startOfDay(for: Date())

        if let nDays = range.days {
            startDate = cal.date(byAdding: .day, value: -(nDays - 1), to: endDate)!
        } else if let earliest = spend.map({ $0.day }).min() {
            startDate = earliest
        } else {
            startDate = endDate
        }

        var cur = startDate
        let spendMap = Dictionary(uniqueKeysWithValues: spend.map { ($0.date, $0) })

        while cur <= endDate {
            let curKey = dayFormatter.string(from: cur)
            if let existing = spendMap[curKey] {
                filledSpend.append(existing)
            } else {
                filledSpend.append(DaySpend(date: curKey, day: cur, cost: 0.0))
            }
            guard let next = cal.date(byAdding: .day, value: 1, to: cur) else { break }
            cur = next
        }

        s.dailySpend = filledSpend

        return s
    }

    nonisolated static func prettyModel(_ id: String) -> String {
        var n = id
        n = n.replacingOccurrences(of: "claude-", with: "Claude ")
        n = n.replacingOccurrences(of: "gpt-", with: "GPT-")
        // strip date suffixes like -20251101
        if let r = n.range(of: #"-\d{8}$"#, options: .regularExpression) {
            n.removeSubrange(r)
        }
        n = n.replacingOccurrences(of: "opus", with: "Opus")
        n = n.replacingOccurrences(of: "sonnet", with: "Sonnet")
        n = n.replacingOccurrences(of: "haiku", with: "Haiku")
        return n
    }

    nonisolated static func modelColor(_ id: String) -> Color {
        if id.contains("opus") { return Color(hex: 0xD97757) }      // Anthropic warm
        if id.contains("sonnet") { return Color(hex: 0xCC8B5C) }
        if id.contains("haiku") { return Color(hex: 0xE0A971) }
        if id.contains("gpt") { return Color(hex: 0x10A37F) }       // OpenAI green
        if id.contains("gemini") { return Color(hex: 0x4285F4) }
        return Color(hex: 0x8E8E93)
    }

    // MARK: - Date formatting

    nonisolated static let dayFormatter: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = TimeZone.current
        f.dateFormat = "yyyy-MM-dd"
        return f
    }()

    nonisolated(unsafe) static let isoParser: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    nonisolated(unsafe) static let isoParserNoFrac: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    nonisolated static func parseDate(_ s: String) -> Date? {
        if let d = isoParser.date(from: s) { return d }
        return isoParserNoFrac.date(from: s)
    }
}

File: windows/PiStats/Core/StatsEngine.cs
using System.ComponentModel;
using System.Globalization;
using System.Runtime.CompilerServices;
using System.Text.RegularExpressions;
using System.Windows.Media;
using PiStats.Models;

namespace PiStats.Core;

/// <summary>
/// Loads/caches the per-day aggregates and derives range summaries.
/// Port of the Swift StatsEngine (ObservableObject -> INotifyPropertyChanged).
/// </summary>
public sealed partial class StatsEngine : INotifyPropertyChanged
{
    private bool _loading = true;
    private double _progress;
    private string? _error;
    private DateTime? _lastUpdated;

    public bool Loading { get => _loading; private set => Set(ref _loading, value); }
    public double Progress { get => _progress; private set => Set(ref _progress, value); }
    public string? Error { get => _error; private set => Set(ref _error, value); }
    public DateTime? LastUpdated { get => _lastUpdated; private set => Set(ref _lastUpdated, value); }

    private List<DayAgg> _days = new();

    // MARK: - Paths (switch between local and remote-synced data)

    private static bool RemoteEnabled => Services.SettingsStore.Shared.RemoteSyncEnabled;

    public static string SessionsDir =>
        RemoteEnabled ? PiPaths.RemoteSessionsDir : PiPaths.LocalSessionsDir;
    public static string CacheFile =>
        RemoteEnabled ? PiPaths.RemoteCacheFile : PiPaths.LocalCacheFile;

    // MARK: - Loading

    public async Task LoadAsync(bool force = false)
    {
        Loading = true;
        Progress = 0;
        Error = null;

        try
        {
            if (RemoteEnabled)
            {
                try
                {
                    await PerformRemoteSyncAsync();
                }
                catch (Exception ex)
                {
                    // Fall back to cached remote data if the sync fails.
                    if (!System.IO.File.Exists(CacheFile)) throw;
                    System.Diagnostics.Debug.WriteLine($"Remote sync failed: {ex.Message}. Using cache.");
                }
            }

            var sessionsDir = SessionsDir;
            var cacheFile = CacheFile;
            var agg = await Task.Run(() =>
                SessionParser.BuildAggregate(sessionsDir, cacheFile, force,
                    p => Progress = p));

            _days = agg.Days;
            LastUpdated = agg.GeneratedAt;
            Progress = 1;
        }
        catch (Exception ex)
        {
            Error = ex.Message;
        }
        finally
        {
            Loading = false;
        }
    }

    public StatsSummary Summary(TimeRange range) => Summarize(_days, range);

    // MARK: - Remote sync

    public static async Task PerformRemoteSyncAsync()
    {
        var s = Services.SettingsStore.Shared;
        await Services.RemoteSync.SyncAsync(
            host: s.RemoteHost,
            port: s.RemotePort,
            user: s.RemoteUser,
            keyPath: s.RemoteKeyPath,
            remotePath: string.IsNullOrWhiteSpace(s.RemotePath) ? "~/.pi/agent/sessions" : s.RemotePath,
            localPath: PiPaths.RemoteSessionsDir);
    }

    // MARK: - Today metrics (for the tray title)

    private DayAgg? TodayAgg()
    {
        var key = DateTime.Today.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);
        return _days.FirstOrDefault(d => d.Date == key);
    }

    public double TodayCost => TodayAgg()?.Cost ?? 0;
    public int TodayLines => TodayAgg()?.LangLines.Values.Sum() ?? 0;
    public int TodayMessages => TodayAgg() is { } d ? d.UserMsgs + d.AsstMsgs : 0;
    public int TodaySessions => TodayAgg()?.SessionIds.Count ?? 0;
    public double TotalCostAll => _days.Sum(d => d.Cost);

    // MARK: - Summarizing

    public static StatsSummary Summarize(List<DayAgg> days, TimeRange range)
    {
        var todayKey = DateTime.Today.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);

        List<DayAgg> filtered;
        var nDays = range.Days();
        if (nDays is int n)
        {
            var cutoff = DateTime.Today.AddDays(-(n - 1));
            filtered = days.Where(d =>
                DateTime.TryParseExact(d.Date, "yyyy-MM-dd", CultureInfo.InvariantCulture,
                    DateTimeStyles.None, out var dt) && dt >= cutoff).ToList();
        }
        else
        {
            filtered = days;
        }

        var s = new StatsSummary();
        var sessionSet = new HashSet<string>();
        var langLines = new Dictionary<string, int>();
        var langEdits = new Dictionary<string, int>();
        var modelCost = new Dictionary<string, double>();
        var modelCount = new Dictionary<string, int>();
        var projCost = new Dictionary<string, double>();
        var projSessions = new Dictionary<string, HashSet<string>>();
        var toolCount = new Dictionary<string, int>();
        var spend = new List<DaySpend>();

        static void Add<TK>(Dictionary<TK, int> map, TK k, int v) where TK : notnull
            => map[k] = map.GetValueOrDefault(k) + v;
        static void AddD<TK>(Dictionary<TK, double> map, TK k, double v) where TK : notnull
            => map[k] = map.GetValueOrDefault(k) + v;

        foreach (var d in filtered)
        {
            s.TotalCost += d.Cost;
            s.InTok += d.InTok; s.OutTok += d.OutTok;
            s.CrTok += d.CrTok; s.CwTok += d.CwTok;
            s.UserMsgs += d.UserMsgs; s.AsstMsgs += d.AsstMsgs;
            s.ToolResults += d.ToolResults;
            sessionSet.UnionWith(d.SessionIds);
            foreach (var (k, v) in d.LangLines) Add(langLines, k, v);
            foreach (var (k, v) in d.LangEdits) Add(langEdits, k, v);
            foreach (var (k, v) in d.ModelCost) AddD(modelCost, k, v);
            foreach (var (k, v) in d.ModelCount) Add(modelCount, k, v);
            foreach (var (k, v) in d.ProjectCost) AddD(projCost, k, v);
            foreach (var (k, ids) in d.ProjectSessions)
            {
                if (!projSessions.TryGetValue(k, out var set)) projSessions[k] = set = new();
                set.UnionWith(ids);
            }
            foreach (var (k, v) in d.ToolCount) Add(toolCount, k, v);
            if (d.Date == todayKey) s.TodayCost += d.Cost;
            if (DateTime.TryParseExact(d.Date, "yyyy-MM-dd", CultureInfo.InvariantCulture,
                    DateTimeStyles.None, out var dt))
                spend.Add(new DaySpend(d.Date, dt, d.Cost));
        }

        s.SessionCount = sessionSet.Count;
        s.DaysActive = filtered.Count(d => d.Cost > 0 || d.AsstMsgs > 0);
        s.AvgCostPerDay = s.DaysActive > 0 ? s.TotalCost / s.DaysActive : 0;

        // Languages
        s.Languages = langLines.Select(kv =>
        {
            var info = LanguageMap.InfoFor(kv.Key);
            return new LangStat(kv.Key, kv.Value, langEdits.GetValueOrDefault(kv.Key),
                info.Color, info.Symbol);
        })
        .Where(l => l.Edits > 0)
        .OrderByDescending(l => l.Lines).ThenByDescending(l => l.Edits)
        .ToList();

        // Models
        s.Models = modelCount.Select(kv =>
            new ModelStat(kv.Key, PrettyModel(kv.Key), modelCost.GetValueOrDefault(kv.Key),
                kv.Value, ModelColor(kv.Key)))
            .OrderByDescending(m => m.Cost)
            .ToList();

        // Projects
        s.Projects = projCost.Select(kv =>
            new ProjectStat(kv.Key, kv.Value, projSessions.GetValueOrDefault(kv.Key)?.Count ?? 0))
            .OrderByDescending(p => p.Cost)
            .ToList();

        // Tools
        s.Tools = toolCount.Select(kv => new ToolStat(kv.Key, kv.Value))
            .OrderByDescending(t => t.Count)
            .ToList();

        // Daily spend (fill gaps so the chart is continuous)
        var endDate = DateTime.Today;
        DateTime startDate;
        if (nDays is int nn)
            startDate = endDate.AddDays(-(nn - 1));
        else if (spend.Count > 0)
            startDate = spend.Min(x => x.Day);
        else
            startDate = endDate;

        var spendMap = spend
            .GroupBy(x => x.Date)
            .ToDictionary(g => g.Key, g => g.First());

        var filledSpend = new List<DaySpend>();
        for (var cur = startDate; cur <= endDate; cur = cur.AddDays(1))
        {
            var key = cur.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);
            filledSpend.Add(spendMap.TryGetValue(key, out var ex)
                ? ex
                : new DaySpend(key, cur, 0.0));
        }
        s.DailySpend = filledSpend;

        return s;
    }

    // MARK: - Model display helpers

    public static string PrettyModel(string id)
    {
        var n = id.Replace("claude-", "Claude ").Replace("gpt-", "GPT-");
        n = DateSuffix().Replace(n, "");
        n = n.Replace("opus", "Opus").Replace("sonnet", "Sonnet").Replace("haiku", "Haiku");
        return n;
    }

    public static Color ModelColor(string id)
    {
        if (id.Contains("opus")) return LanguageMap.Hex(0xD97757);
        if (id.Contains("sonnet")) return LanguageMap.Hex(0xCC8B5C);
        if (id.Contains("haiku")) return LanguageMap.Hex(0xE0A971);
        if (id.Contains("gpt")) return LanguageMap.Hex(0x10A37F);
        if (id.Contains("gemini")) return LanguageMap.Hex(0x4285F4);
        return LanguageMap.Hex(0x8E8E93);
    }

    [GeneratedRegex(@"-\d{8}$")]
    private static partial Regex DateSuffix();

    // MARK: - INotifyPropertyChanged

    public event PropertyChangedEventHandler? PropertyChanged;

    private void Set<T>(ref T field, T value, [CallerMemberName] string? name = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value)) return;
        field = value;
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
    }
}
